File: batch_uploads_imageuploader.pl
#!/usr/bin/perl -w

=pod

=head1 NAME

batch_uploads_imageuploader -- a script that runs C<imaging_upload_file.pl> in
batch mode

=head1 SYNOPSIS

./batch_uploads_imageuploader -profile prod < list_of_scans.txt > log_batch_imageuploader.txt 2>&1 C[options]>

Available options are:

-profile: name of the config file in C<../dicom-archive/.loris_mri>

-verbose: if set, be verbose


=head1 DESCRIPTION


This script runs the Loris-MRI insertion pipeline on multiple scans. The list of
scans are provided through a text file (e.g. C<list_of_scans.txt>) with one scan
details per line.
The scan details includes the path to the scan, identification as to whether the
scan is for a phantom (Y) or not (N), and the candidate name for non-phantom
entries.

Like the LORIS Imaging Uploader interface, this script also validates the
candidate's name against the (start of the) filename and creates an entry in the
C<mri_upload> table.

An example of what C<list_of_scans.txt> might contain for 3 uploads to be
inserted:

 /data/incoming/PSC0001_123457_V1.tar.gz N PSC0000_123456_V1
 /data/incoming/lego_Phantom_MNI_20140101.zip Y
 /data/incoming/PSC0001_123457_V1_RES.tar.gz N PSC0000_123456_V1


=head2 Methods

=cut


use strict;
use warnings;
no warnings 'once';
use File::Basename;
use Getopt::Tabular;
use NeuroDB::DBI;
use NeuroDB::Notify;
use NeuroDB::ExitCodes;

use NeuroDB::Database;
use NeuroDB::DatabaseException;

use NeuroDB::objectBroker::ObjectBrokerException;
use NeuroDB::objectBroker::ConfigOB;



my $profile   = '';
my $upload_id = undef; 
my ($debug, $verbose) = (0,1);
my $stdout = '';
my $stderr = '';

my @opt_table           = (
    [ "Basic options", "section" ],
    [
        "-profile", "string", 1, \$profile,
        "name of config file in ../dicom-archive/.loris_mri"
    ],
    ["-verbose", "boolean", 1,   \$verbose, "Be verbose."]
);

my $Help = <<HELP;
******************************************************************************
Run imaging_upload_file.pl in batch mode
******************************************************************************

This script runs the Loris-MRI insertion pipeline on multiple scans. The list of
scans are provided through a text file (e.g. C<list_of_scans.txt>) with one scan
details per line.
The scan details includes the path to the scan, identification as to whether the
scan is for a phantom (Y) or not (N), and the candidate name for non-phantom
entries.

Like the LORIS Imaging Uploader interface, this script also validates the
candidate's name against the (start of the) filename and creates an entry in the
mri_upload table.

An example of what C<list_of_scans.txt> might contain for 3 uploads to be
inserted:

 /data/incoming/PSC0001_123457_V1.tar.gz N PSC0000_123456_V1
 /data/incoming/Lego_Phantom_MNI_20140101.zip Y
 /data/incoming/PSC0001_123457_V1_RES.tar.gz N PSC0000_123456_V1

Documentation: perldoc batch_uploads_imageuploader

HELP
my $Usage = <<USAGE;
usage: ./batch_uploads_imageuploader.pl -profile prod < list_of_scans.txt > log_batch_imageuploader.txt 2>&1 [options]
       $0 -help to list options
USAGE
&Getopt::Tabular::SetHelp( $Help, $Usage );
&Getopt::Tabular::GetOptions( \@opt_table, \@ARGV )
    || exit $NeuroDB::ExitCodes::GETOPT_FAILURE;

################################################################
################ Get config setting#############################
################################################################
if ( !$profile ) {
    print $Help;
    print STDERR "$Usage\n\tERROR: missing -profile argument\n\n";
    exit $NeuroDB::ExitCodes::PROFILE_FAILURE;
}
{ package Settings; do "$ENV{LORIS_CONFIG}/.loris_mri/$profile" }
if ( !@Settings::db ) {
    print STDERR "\n\tERROR: You don't have a \@db setting in the file "
                 . "$ENV{LORIS_CONFIG}/.loris_mri/$profile \n\n";
    exit $NeuroDB::ExitCodes::DB_SETTINGS_FAILURE;
}


# ----------------------------------------------------------------
## Establish database connection
# ----------------------------------------------------------------

# old database connection
my $dbh = &NeuroDB::DBI::connect_to_db(@Settings::db);

# new Moose database connection
my $db  = NeuroDB::Database->new(
    databaseName => $Settings::db[0],
    userName     => $Settings::db[1],
    password     => $Settings::db[2],
    hostName     => $Settings::db[3]
);
$db->connect();


# ----------------------------------------------------------------
## Get config setting using ConfigOB
# ----------------------------------------------------------------

my $configOB = NeuroDB::objectBroker::ConfigOB->new(db => $db);

my $data_dir  = $configOB->getDataDirPath();
my $mail_user = $configOB->getMailUser();
my $bin_dir   = $configOB->getMriCodePath();
my $is_qsub   = $configOB->getIsQsub();




my ($stdoutbase, $stderrbase) = ("$data_dir/batch_output/imuploadstdout.log", 
				 "$data_dir/batch_output/imuploadstderr.log");

while($_ = $ARGV[0] // '', /^-/) {
    shift;
    last if /^--$/; ## -- ends argument processing
    if (/^-D/) { $debug++ } ## debug level
    if (/^-v/) { $verbose++ } ## verbosity
}

## read input from STDIN, store into array @inputs (`find ....... | this_script`)
my @patientnamearray = ();
my @fullpatharray = ();
my @phantomarray = ();
my @submitted = ();

my $counter = 0;

while(my $line = <STDIN>)
{
    chomp $line;

    my @linearray = split(" " , $line);
    push (@fullpatharray,    $linearray[0]);
    push (@phantomarray,     $linearray[1]);
    push (@patientnamearray, $linearray[2]);
}
close STDIN;

## foreach series, batch magic
foreach my $input (@fullpatharray)
{
    $counter++;
    $stdout = $stdoutbase.$counter;
    $stderr = $stderrbase.$counter;

    #$stdout = '/dev/null';
    #$stderr = '/dev/null';

    my $fullpath    = $fullpatharray[$counter-1];
    my $phantom     = $phantomarray[$counter-1];
    my $patientname = $patientnamearray[$counter-1];

    ## Ensure that 
    ## 1) the uploaded file is of type .tgz or .tar.gz or .zip
    ## 2) check that input file provides phantom details (Y for phantom, N for real candidates)
    ## 3) for non-phantoms, the patient name and path entries are identical; this mimics the imaging uploader in the front-end
    my ($base,$path,$type) = fileparse($fullpath, qr{\..*});
    if (($type ne '.tgz') && ($type ne '.tar.gz') && ($type ne '.zip')) {
	print STDERR "The file on line $counter is not of type .tgz, tar.gz, or "
                 . ".zip and will not be processed\n";
	exit $NeuroDB::ExitCodes::FILE_TYPE_CHECK_FAILURE;
    }
    if (($phantom eq '') || (($phantom ne 'N') && ($phantom ne 'Y'))) {
	print STDERR "Make sure the Phantom entry is filled out "
	             . "with Y if the scan if for a phantom, and N otherwise\n";
	exit $NeuroDB::ExitCodes::PHANTOM_ENTRY_FAILURE;
    }
    if ($phantom eq 'N') {
        if ($patientname ne (substr ($base, 0, length($patientname)))) {
       	    print STDERR "Make sure the patient name $patientname for "
	                     . "non-phantom entries matches the start of $base "
	                     . "filename in $path\n";
	        exit $NeuroDB::ExitCodes::PNAME_FILENAME_MISMATCH;
	    }
    }
    else {
        if ($patientname ne '') {
       	    print STDERR "Please leave the patient name blank for phantom "
       	                 . "entries\n";
	        exit $NeuroDB::ExitCodes::PNAME_FILENAME_MISMATCH;
	}
	else {
	    $patientname = 'NULL';
	}
    }

    ## Populate the mri_upload table with necessary entries and get an upload_id 

    $upload_id = insertIntoMRIUpload(\$dbh,
				     $patientname,
                                     $phantom,
                                     $fullpath);

    ## this is where the subprocesses are created...  should basically run processor script with study directory as argument.
    ## processor will do all the real magic

    my $command = "$bin_dir/uploadNeuroDB/imaging_upload_file.pl "
		. "-profile $profile -upload_id $upload_id $fullpath";
    if ($verbose) {
        $command .= " -verbose";
    }

    ##if qsub is enabled use it
    if ($is_qsub) {
	     open QSUB, "| qsub -V -e $stderr -o $stdout -N process_imageuploader_${counter}";
    	 print QSUB $command;
    	 close QSUB;
    }
    ##if qsub is not enabled
    else {
	print "Running now the following command: $command\n" if $verbose;
	system($command);
    }

     push @submitted, $input;
}
open MAIL, "|mail $mail_user";
print MAIL "Subject: BATCH_UPLOADS_IMAGEUPLOADER: ".scalar(@submitted)." studies submitted.\n";
print MAIL join("\n", @submitted)."\n";
close MAIL;


################################################################
############### insertIntoMRIUpload ############################
################################################################

=pod
insertIntoMRIUpload()
Description:
  - Insert into the mri_upload table entries for data coming
    from batch_upload_imageuploader.pl

=head3 insertIntoMRIUpload($patientname, $phantom, $fullpath)

Function that inserts into the C<mri_upload> table entries for data coming from
the list of scans in the text file provided when calling
C<batch_upload_imageuploader>

INPUTS:
    - $patientname  : The patient name
    - $phantom      : 'Y' if the entry is for a phantom,
                      'N' otherwise
    - $fullpath     : Path to the uploaded file



RETURNS: $upload_id : The upload ID

=cut



sub insertIntoMRIUpload {

    my ( $dbhr, $patientname, $phantom, $fullpath ) = @_;
    my $User = getpwuid($>);

    my $query = "INSERT INTO mri_upload ".
                "(UploadedBy, UploadDate, PatientName, ".
                "IsPhantom, UploadLocation) ".
                "VALUES (?, now(), ?, ?, ?)";
    my $mri_upload_insert = $dbh->prepare($query);
    $mri_upload_insert->execute($User,$patientname,
                                $phantom, $fullpath);

    my $where = " WHERE mu.UploadLocation =?";
    $query = "SELECT mu.UploadID FROM mri_upload mu";
    $query .= $where;
    my $sth = $dbh->prepare($query);
    $sth->execute($fullpath);
    my $upload_id = $sth->fetchrow_array;

    return $upload_id;
}

## exit $NeuroDB::ExitCodes::SUCCESS for find to consider this -cmd true (in case we ever run it that way...)
exit $NeuroDB::ExitCodes::SUCCESS;

__END__

=pod

=head1 LICENSING

License: GPLv3

=head1 AUTHORS

LORIS community <loris.info@mcin.ca> and McGill Centre for Integrative
Neuroscience

=cut

File: batch_uploads_imageuploader.py
import argparse
import textwrap
import sys, getopt  # For script options
import os           # Operating System library to create directories and files
import errno        # For python 2.7 compatibility
import getpass      # For input prompt not to show what is entered
import json         # Provide convenient functions to handle JSON objects 
import requests     # To handle HTTP requests

import warnings
warnings.simplefilter('ignore') # Because I am using unverified ssl certificates 

parser = argparse.ArgumentParser(
    formatter_class=argparse.RawDescriptionHelpFormatter,
    description=textwrap.dedent("""\
This script runs the Loris-MRI insertion pipeline on multiple scans. The list of
scans are provided through a text file (e.g. C<list_of_scans.txt>) with one scan
details per line.

The scan details includes the path to the scan, identification as to whether the
scan is for a phantom (Y) or not (N), and the candidate name. As opposed to the
perl script, this one always requires a patient_name, even for non-phantom entries.

Like the LORIS Imaging Uploader interface, this script also validates the
candidate's name against the (start of the) filename and creates an entry in the
C<mri_upload> table.

An example of what C<list_of_scans.txt> might contain for 3 uploads to be
inserted:

 /data/incoming/PSC0001_123457_V1.tar.gz N PSC0000_123456_V1
 /data/incoming/lego_Phantom_MNI_20140101.zip Y
 /data/incoming/PSC0001_123457_V1_RES.tar.gz N PSC0000_123456_V1

To start the mri_upload process on the LORIS server for each file, use the 
autostart (-a) option when calling this script.

"""
))

parser.add_argument('-b', '--baseurl', required=True, help='LORIS API url with version.')
parser.add_argument('-i', '--inputfile', required=True, help='The list of scans.')
parser.add_argument('-a', '--autostart', action='store_const', const=True, help='If present, will attempt to start the mri_upload process for each scans.')

args = parser.parse_args()

def prettyPrint(string):
    print(json.dumps(string, indent=2, sort_keys=True))

def dicomUpload(filepath, isphantom, patientname):
    pscid, candid, visit_label = patientname
    filename = os.path.basename(filepath)
    files = {'mriFile': (filename, open(filepath, "rb"), 'application/x-tar')}
    payload = {
        'IsPhantom': 'true' if isphantom == 'Y' else 'false',
        'CandID': candid,
        'PSCID': pscid,
        'Visit': visit_label
    }
    response = requests.post(
        url = args.baseurl + '/candidates/' + candid + '/' + visit_label + '/dicoms/',
        headers = {'Authorization': 'Bearer %s' % jwtoken, 'LORIS-Overwrite': 'overwrite'},
        verify = False,
        data = payload,
        files = files
    )
    responsebody = response.content.decode('ascii')
    if (response.status_code != 200):
        raise ValueError(responsebody)
    
    
    data = json.loads(responsebody)
    return data['mri_uploads'].pop()['mri_upload_id']

def startProcess(filename, patientname, mri_upload_id):
    pscid, candid, visit_label = patientname
    payload = {
        'ProcessType': 'mri_upload',
        'MRIUploadID': mri_upload_id
    }
    response = requests.post(
        url = args.baseurl + '/candidates/' + candid + '/' + visit_label + '/dicoms/' + filename + '/processes',
        headers = {'Authorization': 'Bearer %s' % jwtoken},
        verify = False,
        data = payload
    )
    if (response.status_code != 202):
        raise ValueError(responsebody)
        
    return json.loads(response.content.decode('ascii'))

if __name__ == '__main__':
    
    credentials = {
        'username': input('LORIS username: '), 
        'password': getpass.getpass('password: ')
    }
    
    jwtoken = json.loads(requests.post(
        url = args.baseurl + '/login',
        json = credentials,
        verify = False
    ).content.decode('ascii'))['token']
    
    for line in open(args.inputfile):
        if (len(line) == 0):
           continue 

        params      = line.split(' ')
        if (len(params) != 3):
            print('Invalid line: ' + line)
            sys.exit(1)
        filepath    = params[0]
        isphantom   = params[1]
        patientname = params[2].strip()
        
        try:
            open(filepath, 'rb')
        except FileNotFoundError as err:
            print('Skipping ' + filepath)
            print(err.args)
            continue
        
        print('Uploading ' + filepath)
        try:
            mri_upload_id = dicomUpload(filepath, isphantom, patientname.split('_',2))
        except ValueError as err:
            print('Upload failed')
            print(err.args[0])
            continue
            
        print('Upload succesfull')
        print('mri_upload_id: ' + str(mri_upload_id))
        
        if (args.autostart):
            try:
                responsebody = startProcess(os.path.basename(filepath), patientname.split('_',2), mri_upload_id)
            except ValueError as err:
                print('Upload failed')
                print(err.args[0])
                continue
                
            print('mri_upload process started')
            prettyPrint(responsebody)
            print('***************************')
            print('')

